data/generate_embeddings.py

```diff
@@ -0,0 +1,598 @@
+
+# Optional dependencies with fallbacks
+try:
+    import tiktoken
+    TIKTOKEN_AVAILABLE = True
+except ImportError:
+    TIKTOKEN_AVAILABLE = False
+
+try:
+    from tenacity import retry, stop_after_attempt, wait_exponential, retry_if_exception_type
+    TENACITY_AVAILABLE = True
+except ImportError:
+    TENACITY_AVAILABLE = False
+
+
+# -------------------------
+# Defaults & Constants
+# -------------------------
+DEFAULT_MODEL = "text-embedding-3-small"
+DEFAULT_PROVIDER = "openai"
+CLEANING_VERSION = "v1.2"
+METADATA_CSV_PATH = "metadata.csv"
+SUPPORTED_EXTENSIONS = {".txt", ".json"}
+DEFAULT_INPUT_DIR = "data/raw"
+DEFAULT_OUTPUT_DIR = "data/embeddings"
+DEFAULT_MAX_TOKENS = 450
+DEFAULT_OVERLAP = 0.2
+DEFAULT_BATCH_SIZE = 64
+DEFAULT_POLITE_DELAY = 0.1
+DEFAULT_RETRY_ATTEMPTS = 5
+DEFAULT_RETRY_MIN_WAIT = 1
+DEFAULT_RETRY_MAX_WAIT = 30
+
+
+
+# -------------------------
+# Text cleaning & tokenization
+# -------------------------
+def basic_clean(text: str) -> str:
+    """Clean and normalize text content."""
+    if not text:
+        return ""
+    # Normalize whitespace
+    text = re.sub(r"[\r\t]", " ", text)
+    text = re.sub(r"\u00A0", " ", text)
+    # Collapse multiple spaces/newlines
+    text = re.sub(r"\s+", " ", text).strip()
+    # Remove non-informative symbols (keep punctuation, letters, numbers)
+    text = re.sub(r"[^\w\s\.,;:!\?\-\(\)/]", "", text)
+    return text
+
+
+def get_encoder():
+    """Get tiktoken encoder if available."""
+    if not TIKTOKEN_AVAILABLE:
+        return None
+    try:
+        return tiktoken.get_encoding("cl100k_base")
+    except Exception as e:
+        print(f"Error initializing tiktoken encoder: {e}")
+        return None
+
+
+def tokenize_text(text: str, encoder) -> List[int]:
+    """Tokenize text using encoder or fallback to word-based."""
+    if encoder is None:
+        # Fallback: split by words
+        return text.split()
+    return encoder.encode(text)
+
+
+def detokenize_tokens(tokens: List[Any], encoder) -> str:
+    """Convert tokens back to text."""
+    if encoder is None:
+        if isinstance(tokens, list):
+            return " ".join(tokens)
+        return str(tokens)
+    return encoder.decode(tokens)
+
+
+def sliding_window_chunk_tokens(
+    tokens: List[Any], 
+    max_tokens: int, 
+    overlap_ratio: float
+) -> List[List[Any]]:
+    """Create overlapping chunks using sliding window approach."""
+    if max_tokens <= 0:
+        return []
+    if not tokens:
+        return []
+    
+    overlap = int(max_tokens * overlap_ratio)
+    overlap = min(overlap, max_tokens - 1) if max_tokens > 1 else 0
+
+    chunks: List[List[Any]] = []
+    start = 0
+    
+    while start < len(tokens):
+        end = min(start + max_tokens, len(tokens))
+        chunks.append(tokens[start:end])
+        
+        if end == len(tokens):
+            break
+            
+        start = end - overlap if overlap > 0 else end
+    
+    return chunks
```

Comment:
Document the edge cases (e.g., very short texts, overlap > 0.5).
Add a minimum chunk size guard to avoid generating tiny chunks.